File: leetcode/3_longest_substring_without_repeating_characters.cpp
// https://leetcode.com/problems/longest-substring-without-repeating-characters

class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        map<char, const char *> pos;
        int maxsofar = 0;
        
        const char *p = s.c_str(), *q = p;
        
        while (*q != 0)
        {
            auto palready = pos.find(*q);

            if (palready == pos.end())
            {
                pos[*q] = q;
                ++q;
                maxsofar = max(maxsofar, static_cast<int>(distance(p, q)));
                continue;
            }

            for (; p < palready->second + 1; ++p)
                pos.erase(*p);
        }
        
        return maxsofar;
    }
};

// Faster solution with vector for hash:

class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        vector<const char *> pos(256);
        int maxsofar = 0;
        
        const char *p = s.c_str(), *q = p;
        
        while (*q != 0)
        {
            const char *palready = pos[*q];

            if (palready == nullptr)
            {
                pos[*q] = q;
                ++q;
                maxsofar = max(maxsofar, static_cast<int>(distance(p, q)));
                continue;
            }

            for (; p < palready + 1; ++p)
                pos[*p] = nullptr;
        }
        
        return maxsofar;
    }
};

File: leetcode/3_longest_substring_without_repeating_characters.cs
// https://leetcode.com/problems/longest-substring-without-repeating-characters

public class Solution {
    public int LengthOfLongestSubstring(string s) {
        if (s.Length <= 1) return s.Length;

        var lastPos = new Dictionary<char, int>{{ s[0], 0 }};
        int curStart = 0;
        int maxSoFar = 1;

        for (int i = 1; i < s.Length; ++i)
        {
            if (lastPos.TryGetValue(s[i], out int lp) && lp >= curStart)
            {
                maxSoFar = Math.Max(maxSoFar, i - curStart);
                curStart = lp + 1;
            }

            lastPos[s[i]] = i;
        }

        return Math.Max(maxSoFar, s.Length - curStart);
    }
}
